File: packages/crew/src/bus-in-memory.ts
import { randomUUID } from "crypto";
import type { BusMessage, MessageBus, MessageHandler } from "./message-bus.js";

/**
 * In-memory message bus: agents subscribe to channels and receive published messages.
 */
export class InMemoryMessageBus implements MessageBus {
  private handlers = new Map<string, Set<MessageHandler>>();

  async publish(message: Omit<BusMessage, "timestamp" | "messageId">): Promise<void> {
    const full: BusMessage = {
      ...message,
      timestamp: Date.now() / 1000,
      messageId: randomUUID(),
    };
    const set = this.handlers.get(message.channel);
    if (set) {
      for (const h of set) {
        await Promise.resolve(h(full));
      }
    }
  }

  subscribe(channel: string, handler: MessageHandler): () => void {
    if (!this.handlers.has(channel)) {
      this.handlers.set(channel, new Set());
    }
    this.handlers.get(channel)!.add(handler);
    return () => {
      this.handlers.get(channel)?.delete(handler);
    };
  }
}

File: python/src/agent_framework/crew/bus_in_memory.py
"""In-memory message bus."""

import uuid
import time
from collections.abc import Callable
from agent_framework.crew.message_bus import MessageBus, BusMessage, MessageHandler


class InMemoryMessageBus:
    """In-memory message bus."""

    def __init__(self) -> None:
        self._handlers: dict[str, set[MessageHandler]] = {}

    async def publish(self, channel: str, sender_id: str, payload: object) -> None:
        msg = BusMessage(
            channel=channel,
            sender_id=sender_id,
            payload=payload,
            timestamp=time.time(),
            message_id=str(uuid.uuid4()),
        )
        handlers = self._handlers.get(channel, set())
        for h in handlers:
            result = h(msg)
            if hasattr(result, "__await__"):
                await result

    def subscribe(self, channel: str, handler: MessageHandler) -> Callable[[], None]:
        if channel not in self._handlers:
            self._handlers[channel] = set()
        self._handlers[channel].add(handler)

        def unsubscribe() -> None:
            self._handlers.get(channel, set()).discard(handler)

        return unsubscribe
